Make CacheTagArray great again (#221, #225)

denislos opened the pull request:
I have created BPUCache.h and LRUTagCache.h to implement BPU cache. I have also created some unit tests for it. Actually, there are still tests (`read_no_touch` test, `write` test, etc) to be created because I have added virtually only those ones which are tailored to the `LRUTagCache`.

denislos committed:
Add bpucache

denislos committed:
Add BPUCache and LRUTagCache

denislos committed:
Introduce BPUCache to BPU

denislos committed:
Fix problem with clang and msvc

pavelkryukov commented:
Why did you remove CacheTagArray?? We need that class for dozens of caches we will implement later.

denislos commented:
I made a mistake when I renamed it. However, I think it's the way CacheTagArray should look like when hash tables are used. I dropped "bits and bytes" things because I thought it would be better to work with entries in the cache. I haven't changed the logic.

   So this is how BPUCache ( my CacheTagArray) works. In my opinion, It's also the way original CacheTagArray works.
   Each `set` is an array of `ways`. We count sets from 0 to `number_of_sets - 1` and we also count ways from 0 to `number_of_ways - 1`. When a new element is about to be added, its `set` is evaluated in dependence on the address we pass as an argument to `write` method, but its `way` is the way we decide to use in accordance with the LRU algorithm which is implemented independently for each set.

pavelkryukov commented on simulator/bpu/bpu.h:
```diff
@@ -162,7 +146,7 @@ class BPFactory {
         }
 
-        return map.at( name)->create( size_in_entries, ways, branch_ip_size_in_bits);
+        return map.at( name)->create( size_in_entries, ways);
```
How should we configure # of bits for IP now?

pavelkryukov commented:
> I dropped "bits and bytes" things

What do you mean by "bits and bytes"?

> I thought

Please, contact me first before making "revolutionary" changes. For the most of cases the things are made here due to a kind of "strategic" plans. I have some experience in writing simulators and I know already that something would be wrong ideas. Cache-which-come-together-with-data is one of them.

> it would be better to work with entries in the cache

For many purposes we do not need to model entries in the cache, we may have only tags (for example, data cache — the data is actually taken 

> Each set is an array of ways. We count sets from 0 to number_of_sets - 1 and we also count ways from 0 to number_of_ways - 1. When a new element is about to be added, its set is evaluated in dependence on the address we pass as an argument to write method, but its way is the way we decide to use in accordance with the LRU algorithm which is implemented independently for each set.

Right, that is how cache works.

pavelkryukov commented on simulator/infra/cache/t/miss_rate_sim.cpp:
```diff
@@ -1,115 +0,0 @@
-    /* Same as previous for full-associative cache. */
-    for ( auto cache_size : cache_sizes)
-    {
-        CacheTagArray cta( 1024 * cache_size, 1024 * cache_size / 4);
```
That was deleted with nothing in return.

denislos replied:
I have said in the pull request description that there are some tests to be created so miss_rate_sim converted to GTest is one of them. I have planned to use it as the main and the only test for key methods like `write` and `read_no_touch`. However, I thought it would be enough to have bpu tests and LRUTagCache tests in the initial version.

denislos commented:
By the "bits and bytes" things I mean arguments like cache_size_in_bits and addr_size_in_bits which are not needed if the size in entries is used. I have removed branch_ip_size_in_bits for the same reason. What do you mean by cache-which-comes-together-with-data?
Sorry for not asking you first, I thought it would much better if we had more abstract CacheTagArray class, however, I still don't get why the other approach is better in our situation because we do not use explicitly any of them now (e.g. branch_ip_size_in_bits) ( we use default parameters). If we are not ok with default cell structure of CacheTagArray in the future and do not benefit from its simplicity, we still will be able to convert every structure to it by ourselves. Sorry for this.

denislos committed:
Undo changes in infrastucture and return miss_rate_sim

denislos committed:
Merge branch 'master' of https://github.com/MIPT-ILab/mipt-mips into lru

pavelkryukov commented on simulator/bpu/bpu.h:
```diff
@@ -43,42 +43,33 @@ template<typename T>
 {
     std::vector<std::vector<T>> data;
-    CacheTagArray tags;
+    CacheTagArray cache;
```
Then it should be 'tags'

pavelkryukov commented on simulator/bpu/bpu.h:
```diff
@@ -43,42 +43,33 @@ template<typename T>
-        { }
+    BP( uint32 size_in_entries, uint32 ways)
+        : data( ways, std::vector<T>( size_in_entries / ways))
+        , cache( size_in_entries, ways)
```
Please restore 'branch_ip_size_in_bits'

pavelkryukov commented on simulator/bpu/bpu.h:
```diff
@@ -43,42 +43,33 @@ template<typename T>
-              // we're reusing existing CacheTagArray functionality,
-              // but here we don't split memory in blocks, storing
-              // IP's only, so hardcoding here the granularity of 4 bytes:
-              4,
```
Where did that comment go?

pavelkryukov replied:
What happened to "4"? That line size is used for BPU, but not for data caches.

pavelkryukov commented on simulator/infra/cache/LRUTagCache.h:
```diff
@@ -0,0 +1,92 @@
+        std::unordered_map<Key, typename std::list<Key>::const_iterator> lru_hash{};
+
+        size_type number_of_elements = 0u;
+        size_type CAPACITY;
```
Is there any use case to set "size_type" to anything other than "size_t"?

denislos replied:
It looks better for me not to convert size_t to uint32 and introduce size_type.

pavelkryukov replied:
I think uint32 should be used, as we will never have a cache with # of sets or ways exceeding the 32 bit.

pavelkryukov commented on simulator/infra/cache/cache_tag_array.cpp:
```diff
@@ -1,174 +1,73 @@
-}
 
-std::pair<bool, uint32> CacheTagArray::read( Addr addr)
+std::pair<Addr, uint32> CacheTagArray::read( Addr addr)
```
What does Addr mean here?  There are only two results possible: cache hit (true) and cache miss (false).

denislos replied:
Sorry, It's definitely a mistake

pavelkryukov commented:
> By the "bits and bytes" things I mean arguments like cache_size_in_bits and addr_size_in_bits which are not needed if the size in entries is used.

They are still needed as real HW does aliasing to reduce HW cost. For example, BP may use only 16 LSB of IP. Sometimes it makes a collision if there are two instructions with same 16 LSB, but it happens rarely while much HW is reduced.

pavelkryukov commented:
> we do not use explicitly any of them now

It is not the reason to remove anything. There are a lot of stubs which may be unused at the moment, but they are required on the next iteration of development. If I have to remove something, I’ll remove it by myself or create a new task for student. If you think that something should be removed, please consult me/code owner first, as there might be a reason to keep it.

pavelkryukov commented:
> it would much better if we had more abstract CacheTagArray 

We had it as much abstract as possible, excepting the LRU policy. By adding entries to the CacheTagArray, it becomes a Cache class which has less flexibility (you have to handle the data)

pavelkryukov commented:
> What do you mean by cache-which-comes-together-with-data?

To keep entries (BP entries, for instance) and cache tags in the same class. As I explained, for many cases we can have cache tags only.

denislos committed:
Restore dropped arguments and make improvements

denislos committed:
Merge branch 'master' of https://github.com/MIPT-ILab/mipt-mips into lru

denislos committed:
Fix problem with msvc

denislos committed:
Fix comment

pavelkryukov commented on simulator/infra/cache/cache_tag_array.h:
```diff
@@ -1,82 +1,58 @@
+                       uint32 size_of_line = 4,
+                       uint32 addr_size_in_bits = 32)
+            : Log( false)
+            , number_of_sets( check_arguments( size_in_bytes, 
```
Could you please explain the reason why you removed CacheTagArrayCheck class?

denislos replied:
Why do we need the separate class just to check arguments?

pavelkryukov replied:
Why not? C++ guarantees that ctor of CacheTagArrayCheck class is executed before anything else, so if arguments are ill-formed, we won't go inside the function and get any undefined behavior. Additionally, if an alternative implementation of CacheTagArray existed (some unusual cache), programmer would inherit from CacheTagArrayCheck and explicitly get all the checks. Instead, your function "check_arguments" may be easily dropped by accident, plus it has very unclear return value (I would expect void or Boolean `true` if everything correct at least).

pavelkryukov commented on simulator/infra/cache/cache_tag_array.h:
```diff
@@ -1,82 +1,58 @@
         std::pair<bool, uint32> read( Addr addr);
-        /* find in the cache but do not update LRU info */
         std::pair<bool, uint32> read_no_touch( Addr addr) const;
-        /* create new entry in cache */
```
Deleting valid comments is a sabotage.

pavelkryukov commented on simulator/infra/cache/cache_tag_array.h:
```diff
@@ -1,82 +1,58 @@
 #include <infra/log.h>
 
-/* Replacement algorithm modules (LRU). */
-class LRUInfo
```
LRU was intentionally put to the separate class as there are a lot of alternative policies (optimal replacement, pseudo-LRU, MRU, weighted-LRU etc.). So, please put the separate class back, later we will introduce the polymorphic replacement policy selection.

pavelkryukov commented on simulator/infra/cache/cache_tag_array.h:
```diff
@@ -1,82 +1,58 @@
 /**
- * cache_tag_array.h
- * Header for the cache tag array model.
```
Deleting valid comments is a sabotage.

pavelkryukov commented on simulator/infra/cache/cache_tag_array.cpp:
```diff
@@ -1,174 +1,94 @@
-    return addr / line_size;
-}
+    cache[ num_set].update( num_tag);
+    const auto&[ is_hit, value] = cache[ num_set].find( num_tag);
```
Find is counter-intuition. Usually when we write data to the cache, we should find only a LRU way.

pavelkryukov commented on simulator/infra/cache/t/unit_test.cpp:
```diff
@@ -0,0 +1,166 @@
+    std::pair<bool, uint32> result;
+
+    // try to find an element in the empty cache
+    GTEST_ASSERT_NO_DEATH( result = cache.find( 10););
```
There is no point to have "GTEST_ASSERT_NO_DEATH". If something failed inside, it would crash the test anyway.

denislos replied:
We won't be able to tell what exactly went wrong.

pavelkryukov replied:
In both cases you cannot tell what went wrong without a debugger and/or code review. Wrapping everything into these brackets gives extra ~5% of information, but readability of the code is around zero.

pavelkryukov reviewed (commented):
Let's do integration step by step. Could you please resubmit review only with the test?

denislos commented:
Sorry, I don't get it. What do you mean by resubmitting review only with the test? You want me to commit  only with the unit tests, right?

pavelkryukov commented:
Right.

denislos committed:
Merge branch 'master' into lru

denislos committed:
Move LRU to LRUModule class

denislos committed:
Remove LRUTagCache.h

pavelkryukov commented on simulator/infra/cache/cache_tag_array.cpp:
```diff
@@ -1,177 +1,105 @@
-        lru.touch( set_num, way_num); // update LRU info
-    }
+    if ( result.first)
+        lru_module.update( num_set, num_tag); // update LRU if it's a hit
```
`#` of way should be used here, not a tag, right?

denislos replied:
No, we need a tag here.

pavelkryukov replied:
No, we need not. LRU class should know nothing about tags. It has only two basic operations:
* Put some way to the head of the list (== touch)
* Take a way from the tail ( == update == allocate)

denislos replied:
Then there should be a kind of reflection of WAYS to TAGS( e,g, another map which is not good), because we need somehow to know the least recently used tag to erase it from the hash table.

denislos replied:
Well, I will use another map.

pavelkryukov replied:
> Then there should be a kind of reflection of WAYS to TAGS( e,g, another map which is not good), because we need somehow to know the least recently used tag to erase it from the hash table.

Yes, it should be used for CacheTagArray. It is a common operation for cache ("check what data is in way 1 of set 4").

pavelkryukov replied:
> Well, I will use another map.

Note that for PseudoLRU there is no map inside LRU class, it uses a different mechanism. We might use FIFO replacement as well.

denislos replied:
I have used a map inside CacheTagArray.

pavelkryukov commented on simulator/infra/cache/cache_tag_array.h:
```diff
@@ -1,82 +1,191 @@
 
-        uint32 set( Addr addr) const;
-        Addr tag( Addr addr) const;
+        uint32 set( Addr addr) const { return (addr / line_size) & (number_of_sets - 1); }
```
Recently I've added aliasing to these functions, please return it back.

pavelkryukov commented on simulator/infra/cache/cache_tag_array.h:
```diff
@@ -1,82 +1,191 @@
+            : lru_info( number_of_sets, LRUCacheInfo<Addr>( number_of_ways))
+        { }
+
+        std::pair<Addr, uint32> update( uint32 num_set, Addr addr) 
```
There should be two methods:
* Touch (when an existing element is accessed)
* Update (when an new element is added)

denislos committed:
Make improvements

denislos committed:
Avoid using tags in the LRU module and use an extra map in CacheTagArray

pavelkryukov commented on simulator/infra/cache/cache_tag_array.cpp:
```diff
@@ -1,177 +1,147 @@
 {
-    const auto lookup_result = read_no_touch( addr);
-    const auto&[ is_hit, way_num] = lookup_result;
+    auto result = read_no_touch( addr);
```
Could you please use C++17 syntax?
```c++
const auto&[ is_hit, way_num] = lookup_result;
```

pavelkryukov commented on simulator/infra/cache/cache_tag_array.cpp:
```diff
@@ -1,177 +1,147 @@
-    for ( auto it = list.begin(); it != list.end(); ++it)
+    std::size_t return_value = number_of_elements;
+
+    if ( number_of_elements == CAPACITY)
```
I think you may initialize the hash table in ctor and remove `number_of_elements`

denislos committed:
Use C++17 syntax

pavelkryukov commented on simulator/infra/cache/cache_tag_array.h:
```diff
@@ -1,82 +1,137 @@
+        const Addr   addr_mask;
+
+        // maps to convert num_ways to tags
+        std::vector<std::unordered_map<uint32, Addr>> ways_to_tags;
```
It can be a ```std::vector<std::vector<Addr>>```, right?

pavelkryukov commented on simulator/infra/cache/cache_tag_array.h:
```diff
@@ -1,82 +1,137 @@
+                                  size_of_line,
+                                  addr_size_in_bits)
+            , number_of_sets( size_in_bytes / ( ways * size_of_line))
+            , number_of_ways( ways)
```
Why were these fields moved from CacheTagArrayCheck?

denislos replied:
Why do we need them there? I mean, It's strange to have them in the "Check" class.

pavelkryukov replied:
Denis, that strange kind of discussion is repeating over and over. You are the person who did the changes, so I should be the guy who expects the answers which explain your motivation.

We need it here because we may have it here. "CacheTagArrayCheck" has the highest level of abstraction, so if a class was inherited from it we would not think about correctness of size variables anymore. We may add more methods to the base class which operate with sizes only and then use them for some constant expressions or run-time checks etc. I agree that "Check" may be not the best name, but "CacheTagArraySizeBaseWithBoundaryChecks" has some drawbacks either.

There are _a lot_ of things in C++ programs which may look strange, but they can be even the only one possible solution originated from days and months of the code evolution. If they don't contain a bug, it is better not to touch them.

pavelkryukov commented on simulator/infra/cache/cache_tag_array.cpp:
```diff
@@ -1,177 +1,147 @@
-    if ( line_size == 0)
-        serr << "ERROR: Wrong arguments! Line size should be greater than zero"
+    
+    if ( size_of_line == 0)
```
"line size" is a common idiom for caches, please move it back.

denislos committed:
Remove number_of_elements and make improvements

denislos committed:
Remove number_of_elements

denislos committed:
Restore CacheTagArrayCheck

pavelkryukov commented on simulator/infra/cache/cache_tag_array.h:
```diff
@@ -1,82 +1,127 @@
+        const Addr   addr_mask;
+
+        // to convert num_ways to tags
+        std::vector<std::vector<std::pair<bool, Addr>>> ways_to_tags;
```
`std::pair` is not the best idea.
1. It is usually unclear what "first" and "second" mean
2. One day we will need to add one more field to the table. With pair/tuple we shall re-write everything!!

denislos committed:
Use struct instead of std::pair

pavelkryukov committed:
Merge branch 'master' into lru
